src/content/lib/ua_overrider.js

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const {Cc, Ci} = require("chrome");
+
+const ObserverService = Cc["@mozilla.org/observer-service;1"].getService(Ci.nsIObserverService);
+const DefaultUA = Cc["@mozilla.org/network/protocol;1?name=http"].getService(Ci.nsIHttpProtocolHandler).userAgent;
+const eTLDService = Cc["@mozilla.org/network/effective-tld-service;1"].getService(Ci.nsIEffectiveTLDService);
+
+const NS_HTTP_ON_USERAGENT_REQUEST_TOPIC = "http-on-useragent-request";
+
+class UAOverrider {
+  constructor(overrides) {
+    this._overrides = {};
+    this._overrideForURICache = new Map();
+
+    this.initOverrides(overrides);
+  }
+
+  initOverrides(overrides) {
+    for (let override of overrides) {
+      if (!this._overrides[override.baseDomain]) {
+        this._overrides[override.baseDomain] = [];
+      }
+
+      if (!override.uriMatcher) {
+        override.uriMatcher = () => true;
+      }
+
+      this._overrides[override.baseDomain].push(override);
+    }
```

Comment:
I expect there to be more than one override rule per base domain. Imagine overrides for Google Mail and Google maps, both are hosted on `google.com`, but with different `uriMatcher`s. That's why I use an array here.